README.md
# karma-browserstack-launcher

[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/karma-runner/karma-browserstack-launcher)
 [![npm version](https://img.shields.io/npm/v/karma-browserstack-launcher.svg?style=flat-square)](https://www.npmjs.com/package/karma-browserstack-launcher) [![npm downloads](https://img.shields.io/npm/dm/karma-browserstack-launcher.svg?style=flat-square)](https://www.npmjs.com/package/karma-browserstack-launcher)

[![Build Status](https://img.shields.io/travis/karma-runner/karma-browserstack-launcher/master.svg?style=flat-square)](https://travis-ci.org/karma-runner/karma-browserstack-launcher) [![Dependency Status](https://img.shields.io/david/karma-runner/karma-browserstack-launcher.svg?style=flat-square)](https://david-dm.org/karma-runner/karma-browserstack-launcher) [![devDependency Status](https://img.shields.io/david/dev/karma-runner/karma-browserstack-launcher.svg?style=flat-square)](https://david-dm.org/karma-runner/karma-browserstack-launcher#info=devDependencies)

> Use any browser on [BrowserStack](https://www.browserstack.com/)!


## Installation

The easiest way is to keep `karma-browserstack-launcher` as a devDependency in your `package.json`. Just run,

```bash
$ npm install karma-browserstack-launcher --save-dev
```

and it will be added automatically.


## Configuration

```js
// karma.conf.js
module.exports = function(config) {
  config.set({
    // global config of your BrowserStack account
    browserStack: {
      username: 'jamesbond',
      accessKey: '007'
    },

    // define browsers
    customLaunchers: {
      bs_firefox_mac: {
        base: 'BrowserStack',
        browser: 'firefox',
        browser_version: '21.0',
        os: 'OS X',
        os_version: 'Mountain Lion'
      },
      bs_iphone5: {
        base: 'BrowserStack',
        device: 'iPhone 5',
        os: 'ios',
        os_version: '6.0'
      }
    },

    browsers: ['bs_firefox_mac', 'bs_iphone5']
  })
}
```

### Global options

- `username` your BS username, you can also use `BROWSERSTACK_USERNAME` env variable.
- `accessKey` your BS access key, you can also use `BROWSERSTACK_ACCESS_KEY` env variable.
- `startTunnel` do you wanna establish the BrowserStack tunnel ? (defaults to `true`)
- `tunnelIdentifier` in case you want to start the BrowserStack tunnel outside `karma` by setting `startTunnel` to `false`, set the identifier passed to the `-localIdentifier` option here (optional)
- `retryLimit` how many times do you want to retry to capture the browser ? (defaults to `3`)
- `captureTimeout` the browser capture timeout (defaults to `120`)
- `timeout` the BS worker timeout (defaults to `300`
- `build` the BS worker build name (optional)
- `name` the BS worker name (optional)
- `project` the BS worker project name (optional)
- `proxyHost` the host of your proxy for communicating with BrowserStack REST API and BrowserStackLocal (optional)
- `proxyPort` the port of your proxy (optional)
- `proxyUser` the username used for authentication with your proxy (optional)
- `proxyPass` the password used for authentication with your proxy (optional)
- `proxyProtocol` the protocol of your proxy (optional. default: `http`. valid: `http` or `https`)
- `forcelocal` force traffic through the local BrowserStack tunnel, passes flag through to BrowserStackTunnel
- `video` enable video recording of session on BrowserStack (defaults to `true`)
- `keepAlive` periodically issue a _takeScreenshot_ command to keep long test sessions from timing out (defaults to `false`)
- `keepAliveDelay` the number of milliseconds to wait between issuing keep alive commands (defaults to `60000`ms; 1 minute)

### Per browser options

- `device` name of the device
- `real_mobile` or `realMobile` allows the session to run on a real mobile device instead of an emulator / simulator (optional, defaults to `false`)
- `browser` name of the browser
- `browser_version` version of the browser
- `os` which platform ?
- `os_version` version of the platform
- `build` the BS worker build name (optional, defaults to global)
- `name` the BS worker name (optional, defaults to global)
- `project` the BS worker project name (optional, defaults to global)

> **Note:** you can also pass through any additional options supported by browserstack. (EG. `url`, `resolution`, etc.)  
See https://www.browserstack.com/automate/capabilities for a full list of supported options.

### BrowserStack reporter

To report session results back to BrowserStack for display on your BrowserStack dashboard, use the following additional configuration:

```js
// karma.conf.js
module.exports = function(config) {
  config.set({
    // The rest of your karma config is here
    // ...
    reporters: ['dots', 'BrowserStack']
  })
}
```

### Browserstack iOS simulators

By default, your Selenium and JS tests will run on real iOS devices on BrowserStack. Since we are in the implementation phase, we are still working on a few things, such as adding more devices, ability to test on local URLs, etc.

In case your tests are facing any issues on real iOS devices, we also provide iOS simulators where you can run your automated tests smoothly.

To access our iOS simulators, use the following capabilities:

```js
customLaunchers: {
  iPad_3: {
    real_mobile: false,
    device: 'iPad 3rd (6.0)',
    os: 'ios',
    'os_version': '6.0',
    'browser_version': null,
    browser: 'Mobile Safari'
  }
}
```

List of iOS simulators you can test on:

- `device: 'iPad 3rd', 'os_version': '5.1'`
- `device: 'iPad 3rd (6.0)', 'os_version': '6.0'`
- `device: 'iPad Mini', 'os_version': '7.0'`
- `device: 'iPad 4th', 'os_version': '7.0'`
- `device: 'iPhone 4S', 'os_version': '5.1'`
- `device: 'iPhone 4S (6.0)', 'os_version': '6.0'`
- `device: 'iPhone 5', 'os_version': '6.0'`
- `device: 'iPhone 5S', 'os_version': '7.0'`

### CI/CD Build Environment Variables

Many CI/CD systems will make the name or ID of the currently running build available via an environment variable. The follow environment variables below are supported by default:

* `process.env.BUILD_NUMBER`
* `process.env.BUILD_TAG`
* `process.env.CI_BUILD_NUMBER`
* `process.env.CI_BUILD_TAG`
* `process.env.TRAVIS_BUILD_NUMBER`
* `process.env.CIRCLE_BUILD_NUM`
* `process.env.DRONE_BUILD_NUMBER`

---

[BrowserStack's REST API documentation](https://www.browserstack.com/automate/rest-api#rest-api-browsers)
explains how to retrieve a list of desired capabilities for browsers.

----

For more information on Karma see the [homepage](http://karma-runner.github.io).

----

Check out sample code working with karma-browserstack-launcher [here](https://github.com/browserstack/karma-browserstack-example).


CHANGELOG.md
<a name="1.6.0"></a>
# [1.6.0](https://github.com/karma-runner/karma-browserstack-launcher/compare/v1.5.1...v1.6.0) (2020-01-16)


### Features

* Added new configuration keys allowing for keep alive (via screenshot) of session / worker at a configurable interval 

<a name="1.5.1"></a>
# [1.5.1](https://github.com/karma-runner/karma-browserstack-launcher/compare/v1.3.0...v1.5.1) (2019-04-03)


### Features

* Switched to official browserstack local package, as it is frequently updated and used across multiple products.
* Code refactored. 


<a name="1.4.0"></a>
# [1.4.0](https://github.com/karma-runner/karma-browserstack-launcher/compare/v1.3.0...v1.4.0) (2018-10-27)


### Features

* add support for the enableLoggingForApi property ([b6776f4](https://github.com/karma-runner/karma-browserstack-launcher/commit/b6776f4))
* allow additional browserstack capabitilies to be configured verbatim ([62fc997](https://github.com/karma-runner/karma-browserstack-launcher/commit/62fc997))



<a name="1.3.0"></a>
# [1.3.0](https://github.com/karma-runner/karma-browserstack-launcher/compare/v1.2.0...v1.3.0) (2017-06-28)


### Features

* add support for using BrowserStackReporter with a proxy  ([b0f8031](https://github.com/karma-runner/karma-browserstack-launcher/commit/b0f8031))
* allow the user to disable browserstack video recording ([b3cce6c](https://github.com/karma-runner/karma-browserstack-launcher/commit/b3cce6c))
* **deps:** update to the latest ([6c719d2](https://github.com/karma-runner/karma-browserstack-launcher/commit/6c719d2))
* **deps:** update to tunnel 2.0.1 ([9617438](https://github.com/karma-runner/karma-browserstack-launcher/commit/9617438))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/karma-runner/karma-browserstack-launcher/compare/v1.1.1...v1.2.0) (2017-01-29)


### Features

* pass forcelocal to tunnel ([693a4aa](https://github.com/karma-runner/karma-browserstack-launcher/commit/693a4aa))
* permit overriding `name`, `build`, and `project` on a per-browser basis ([9c77a96](https://github.com/karma-runner/karma-browserstack-launcher/commit/9c77a96))
* Update browserstacktunnel wrapper ([4519a37](https://github.com/karma-runner/karma-browserstack-launcher/commit/4519a37))



<a name="1.1.1"></a>
## [1.1.1](https://github.com/karma-runner/karma-browserstack-launcher/compare/v1.1.0...v1.1.1) (2016-09-05)


### Bug Fixes

* update incorrect launcher require path ([5cc2d9c](https://github.com/karma-runner/karma-browserstack-launcher/commit/5cc2d9c)), closes [#73](https://github.com/karma-runner/karma-browserstack-launcher/issues/73)



<a name="1.1.0"></a>
# [1.1.0](https://github.com/karma-runner/karma-browserstack-launcher/compare/v1.0.1...v1.1.0) (2016-09-05)


### Features

* add BrowserStack reporter ([ddb5779](https://github.com/karma-runner/karma-browserstack-launcher/commit/ddb5779)), closes [#72](https://github.com/karma-runner/karma-browserstack-launcher/issues/72)



<a name="1.0.1"></a>
## [1.0.1](https://github.com/karma-runner/karma-browserstack-launcher/compare/v0.1.11...v1.0.1) (2016-05-31)


### Bug Fixes

* urlencode proxy username and password ([ab42eee](https://github.com/karma-runner/karma-browserstack-launcher/commit/ab42eee))
* Use separate property to shutdown workers ([221d49f](https://github.com/karma-runner/karma-browserstack-launcher/commit/221d49f)), closes [#62](https://github.com/karma-runner/karma-browserstack-launcher/issues/62)

### Features

* Pass proxy url to browserstack api ([0c1ca47](https://github.com/karma-runner/karma-browserstack-launcher/commit/0c1ca47))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/karma-runner/karma-browserstack-launcher/compare/v0.1.11...v1.0.0) (2016-05-04)




<a name="0.1.11"></a>
## [0.1.11](https://github.com/karma-runner/karma-browserstack-launcher/compare/v0.1.10...v0.1.11) (2016-04-09)


### Bug Fixes

* issue with url being 'undefined' after disconnect ([e746ecd](https://github.com/karma-runner/karma-browserstack-launcher/commit/e746ecd)), closes [#63](https://github.com/karma-runner/karma-browserstack-launcher/issues/63)



<a name="0.1.10"></a>
## [0.1.10](https://github.com/karma-runner/karma-browserstack-launcher/compare/v0.1.8...v0.1.10) (2016-02-12)


### Bug Fixes

* Capture timeout occurs even after tests have run successfully ([1bd45b6](https://github.com/karma-runner/karma-browserstack-launcher/commit/1bd45b6)), closes [#17](https://github.com/karma-runner/karma-browserstack-launcher/issues/17)
* Ensure the browser gets terminated ([ea3ed1e](https://github.com/karma-runner/karma-browserstack-launcher/commit/ea3ed1e))

### Features

* add build number env vars for compatibility with drone and other CI systems ([054b8d4](https://github.com/karma-runner/karma-browserstack-launcher/commit/054b8d4))



<a name="0.1.9"></a>
## [0.1.9](https://github.com/karma-runner/karma-browserstack-launcher/compare/v0.1.8...v0.1.9) (2016-01-10)


### Bug Fixes

* Capture timeout occurs even after tests have run successfully ([1bd45b6](https://github.com/karma-runner/karma-browserstack-launcher/commit/1bd45b6)), closes [#17](https://github.com/karma-runner/karma-browserstack-launcher/issues/17)



<a name="0.1.8"></a>
## [0.1.8](https://github.com/karma-runner/karma-browserstack-launcher/compare/v0.1.7...v0.1.8) (2015-12-16)


### Bug Fixes

* Avoid setting tunnelIdentifier if startTunnel is false ([fe61baa](https://github.com/karma-runner/karma-browserstack-launcher/commit/fe61baa))
* Property access on undefined config ([a05e6cf](https://github.com/karma-runner/karma-browserstack-launcher/commit/a05e6cf)), closes [#52](https://github.com/karma-runner/karma-browserstack-launcher/issues/52)

### Features

* **proxy:** Add proxy settings to BrowserstackLocal ([7c3da2b](https://github.com/karma-runner/karma-browserstack-launcher/commit/7c3da2b))



<a name="0.1.6"></a>
## [0.1.6](https://github.com/karma-runner/karma-browserstack-launcher/compare/v0.1.6...v0.1.6) (2015-10-22)




<a name="0.1.5"></a>

index.js
const Q = require('q')
const api = require('browserstack')
const browserstack = require('browserstack-local')
const workerManager = require('./worker-manager')
const BrowserStackReporter = require('./browserstack-reporter')

var createBrowserStackTunnel = function (logger, config, emitter) {
  const log = logger.create('launcher.browserstack')
  const bsConfig = config.browserStack || {}
  if (bsConfig.startTunnel === false) {
    return Q()
  }

  const bsAccesskey = process.env.BROWSERSTACK_ACCESS_KEY || process.env.BROWSER_STACK_ACCESS_KEY || bsConfig.accessKey
  const bsLocal = new browserstack.Local()
  const bsLocalArgs = { key: bsAccesskey }
  const deferred = Q.defer()

  log.debug('Starting BrowserStackLocal')
  bsLocal.start(bsLocalArgs, function () {
    log.debug('Started BrowserStackLocal')
    deferred.resolve()
  })

  emitter.on('exit', function (done) {
    log.debug('Shutting down BrowserStackLocal')
    bsLocal.stop(function () {
      log.debug('Stopped BrowserStackLocal')
      done()
    })
  })

  return deferred.promise
}

var createBrowserStackClient = function (/* config.browserStack */config, /* BrowserStack:sessionMapping */ sessionMapping) {
  var env = process.env

  config = config || {}

  var options = {
    username: env.BROWSERSTACK_USERNAME || env.BROWSER_STACK_USERNAME || config.username,
    password: env.BROWSERSTACK_ACCESS_KEY || env.BROWSER_STACK_ACCESS_KEY || config.accessKey
  }

  if (config.proxyHost && config.proxyPort) {
    config.proxyProtocol = config.proxyProtocol || 'http'
    var proxyAuth = (config.proxyUser && config.proxyPass)
      ? (encodeURIComponent(config.proxyUser) + ':' + encodeURIComponent(config.proxyPass) + '@') : ''
    options.proxy = config.proxyProtocol + '://' + proxyAuth + config.proxyHost + ':' + config.proxyPort
  }

  if (!config.browserstack || config.browserStack.startTunnel !== false) {
    options.Local = true
  }

  sessionMapping.credentials = {
    username: options.username,
    password: options.password,
    proxy: options.proxy
  }

  // TODO(vojta): handle no username/pwd
  var client = api.createClient(options)

  var pollingTimeout = config.pollingTimeout || 1000

  if (!workerManager.isPolling) {
    workerManager.startPolling(client, pollingTimeout, function (err) {
      if (err) {
        console.error(err)
      }
    })
  }

  return client
}

var formatError = function (error) {
  if (error.message === 'Validation Failed') {
    return '  Validation Failed: you probably misconfigured the browser ' +
      'or given browser is not available.'
  }

  return error.toString()
}

var BrowserStackBrowser = function (
  id, emitter, args, logger,
  /* config */ config,
  /* browserStackTunnel */ tunnel,
  /* browserStackClient */ client,
  baseLauncherDecorator,
  captureTimeoutLauncherDecorator,
  retryLauncherDecorator,
  /* BrowserStack:sessionMapping */ sessionMapping
) {
  var self = this

  baseLauncherDecorator(self)
  captureTimeoutLauncherDecorator(self)
  retryLauncherDecorator(self)

  var workerId = null
  var captured = false
  var alreadyKilling = null
  var log = logger.create('launcher.browserstack')
  var browserName = (args.browser || args.device) + (args.browser_version ? ' ' + args.browser_version : '') +
    ' (' + args.os + ' ' + args.os_version + ')'

  this.id = id
  this.name = browserName + ' on BrowserStack'

  var bsConfig = config.browserStack || {}
  var captureTimeout = config.captureTimeout || 0
  var captureTimeoutId
  var retryLimit = bsConfig.retryLimit || 3
  var previousUrl = null
  var keepAliveEnabled = false

  this.start = function (url) {
    url = url || previousUrl
    previousUrl = url

    var globalSettings = Object.assign(
      {
        timeout: 300,
        name: 'Karma test',
        build: process.env.BUILD_NUMBER ||
          process.env.BUILD_TAG ||
          process.env.CI_BUILD_NUMBER ||
          process.env.CI_BUILD_TAG ||
          process.env.TRAVIS_BUILD_NUMBER ||
          process.env.CIRCLE_BUILD_NUM ||
          process.env.DRONE_BUILD_NUMBER || null,
        // TODO(vojta): remove "version" (only for B-C)
        browser_version: args.version || 'latest',
        video: true
      },
      bsConfig
    )

    // TODO(vojta): handle non os/browser/version
    var settings = Object.assign(
      {
        url: url + '?id=' + id,
        'browserstack.tunnel': true
      },
      globalSettings,
      args
    )

    tunnel.then(function () {
      client.createWorker(settings, function (error, worker) {
        var sessionUrlShowed = false

        if (error) {
          log.error('Can not start %s\n  %s', browserName, formatError(error))
          return emitter.emit('browser_process_failure', self)
        }

        workerId = worker.id
        alreadyKilling = null

        worker = workerManager.registerWorker(worker)
        worker.on('status', function (status) {
          // TODO(vojta): show immediately in createClient callback once this gets fixed:
          // https://github.com/browserstack/api/issues/10
          if (!sessionUrlShowed) {
            log.info('%s session at %s', browserName, worker.browser_url)
            sessionMapping[self.id] = worker.browser_url.split('/').slice(-1)[0]
            sessionUrlShowed = true
          }

          switch (status) {
            case 'running':
              keepAliveEnabled = true
              startKeepAlive()
              log.debug('%s job started with id %s', browserName, workerId)

              if (captureTimeout && !captured) {
                captureTimeoutId = setTimeout(self._onTimeout, captureTimeout)
              }
              break

            case 'queue':
              log.debug('%s job with id %s in queue.', browserName, workerId)
              break

            case 'delete':
              keepAliveEnabled = false
              log.debug('%s job with id %s has been deleted.', browserName, workerId)
              break
          }
        })
      })
    }).catch(function () {
      emitter.emit('browser_process_failure', self)
    })
  }

  var startKeepAlive = function () {
    if (bsConfig.keepAlive && keepAliveEnabled) {
      var session = sessionMapping[self.id]
      log.debug(`keepAlive enabled and being executed for ID: '${workerId}, session: '${session}'`)
      try {
        client.takeScreenshot(workerId, function (err, response) {
          if (err) {
            log.error(`unable to issue keep alive due to: ${err}`)
          }
          log.debug(JSON.stringify(response))
        })
      } catch (e) {
        log.error(e)
      }
      var delay = bsConfig.keepAliveDelay || 60000
      setTimeout(startKeepAlive, delay)
    }
  }

  this.kill = function (done) {
    var allDone = function () {
      self._done()
      if (done) {
        done()
      }
    }

    if (!alreadyKilling) {
      alreadyKilling = Q.defer()

      if (workerId) {
        keepAliveEnabled = false
        log.info('Killing %s (worker %s).', browserName, workerId)
        client.terminateWorker(workerId, function () {
          log.debug('%s (worker %s) successfully killed.', browserName, workerId)

          if (captureTimeoutId) {
            clearTimeout(captureTimeoutId)
            captureTimeoutId = null
          }

          workerId = null
          captured = false
          alreadyKilling.resolve()
        })
      } else {
        alreadyKilling.resolve()
      }
    }

    return alreadyKilling.promise.then(allDone)
  }

  this.forceKill = function () {
    var self = this

    return Q.promise(function (resolve) {
      self.kill(resolve)
    })
  }

  this.markCaptured = function () {
    captured = true

    if (captureTimeoutId) {
      clearTimeout(captureTimeoutId)
      captureTimeoutId = null
    }
  }

  this.isCaptured = function () {
    return captured
  }

  this.toString = function () {
    return this.name
  }

  this._onTimeout = function () {
    if (captured) {
      return
    }

    log.warn('%s has not captured in %d ms, killing.', browserName, captureTimeout)
    self.kill(function () {
      if (retryLimit--) {
        self.start(previousUrl)
      } else {
        emitter.emit('browser_process_failure', self)
      }
    })
  }
}

// PUBLISH DI MODULE
module.exports = {
  'browserStackTunnel': ['factory', createBrowserStackTunnel],
  'browserStackClient': ['factory', createBrowserStackClient],
  'launcher:BrowserStack': ['type', BrowserStackBrowser],
  'reporter:BrowserStack': ['type', BrowserStackReporter],
  'BrowserStack:sessionMapping': ['value', {}]
}
